// app/YoutubeSearch.js
var request = require('request'),
    cheerio = require('cheerio'),
    Q = require('q'),
    _ = require('lodash');

module.exports = function youtubeSeach() {

    var that = this;

    that._query = "";
    that._options = {
        maxResults: 10,
        startIndex: 0
    };

    return that;
};

module.exports.prototype.query = function Query(query) {
    if (query)
        this._query = query;
    return this;
};
module.exports.prototype.options = function Options(options) {
    _.assign(this._options, options);
    return this;
};
module.exports.prototype.search = function Search() {
    var that = this,
        deferred = Q.defer();

    var search = "https://www.youtube.com/results?search_query=" + encodeURIComponent(that._query);
    request.get(search, function (err, res, body) {
        if (err) {
            deferred.reject(err);
            return;
        }

        var $ = cheerio.load(body);
        var ret = [];
        $('.section-list .yt-lockup .yt-lockup-title a.yt-uix-tile-link').each(function (index) {
            var url = "http://www.youtube.com" + $(this).attr("href");
            ret.push({
                url: url
            });
        });

        ret = ret.slice(that._options.startIndex, that._options.startIndex + that._options.maxResults);

        if (ret.length <= 0) {
            deferred.reject(new Error("Can't find any video with that query"));
            return;
        }

        deferred.resolve(ret);
    });

    return deferred.promise;
};
